WordPress/gutenberg
# Post excerpt: Add excerpt length control

## What?

Provide a UI control to control the length of the excerpt.
Closes https://github.com/WordPress/gutenberg/issues/34757

Currently there is no way to limit the length of the automatically generated excerpt without code. 
It was requested in discussions in the issue that the feature should limit the length of both the user created and the automatically generated excerpts.

_Description updated January 23._

## Why?

To add the ability to control the excerpt length in a design.

## How?

The PR adds a new block attribute to the post excerpt block called `excerptLength`. The default value is 55 words, to match the WordPress default length for excerpts.

An excerpt length control is added to the block sidebar. The control uses a `RangeControl` component with the min value of 10 and max value of 100.
The control limits the number of words that are displayed by the post excerpt block on the front.
It does **not** limit the number of words that are in the Excerpt panel in the post tab when editing a post.

The limitation is applied to the display of both auto generated and user-generated excerpts.
In index.php, the number of words is limited by passing `get_the_excerpt` and the `excerptLength` attribute through [wp_trim_words](https://developer.wordpress.org/reference/functions/wp_trim_words/).

In the block editor, outside of a query loop:
- When the block is selected, the full text shows and is editable.
- When it is deselected, the trimmed text shows.

In a query loop, the trimmed text shows.

## Testing Instructions
_Please test with both classic and block themes, and with a theme that filters the excerpt length._

1. Create a new post and add a long paragraph.
2. Add a post excerpt block.
3. Open the block settings sidebar. Confirm that there is a range control combined with a numeric input field, with the label "max number of words".
4. Change the value in the control.
5. Deselect the block and confirm that the trimmed text shows, with the number of words matching the value in the control.
5. Save and view the front, confirm that the excerpt displays the correct number of words.

## Testing instructions for screen reader users
Create a new post.
Add a post excerpt block.
Enter a sample excerpt with more than ten words.
Press tab to move to the settings sidebar.
Move past the Posts tab.
In the Block tab, navigate forwards to the Settings panel.
Navigate past the Show link on new line option.
You should now find a new setting for the text length:
Using VoiceOver, I first hear the number of words, the label, and then the type of control:
"55, Max number of words, slider".
There are three ways to change the value:
- Either use the arrow keys on the slider,
- Tab forwards one step and use the arrow keys in the stepper,
- Tab forwards one step from the slider and enter the number into the field in the stepper.

Change the value to 10.
Navigate back to the editor.
Activate the "read all" command or similar for the screen reader.
Confirm that the excerpt only includes the first ten words.
Save and view the post on the front.
Confirm that the excerpt only includes the first ten words.
